docs/getting-started-tutorial/6-adding-usernames.adoc

```diff
@@ -0,0 +1,146 @@
+:idprefix:
+:idseparator: -
+:toc: right
+:toclevels: 2
+
+# Adding Usernames
+
+We're about to add another use case to our chat system:
+
+ - As a chatter, I want to see who posted a message so that I have better
+   context for the conversation.
+
+The first thing we'll do is change the HTML to look like we want it to. Let's
+add the username:
+
+```
+...
+      <li>
+        <span class="poster">Antonio</span> <span class="body">Hi!</span>
+      </li>
+      <li class="clearable">
+        <span class="poster">David</span> <span class="body">Oh, hey there.</span>
+      </li>
+      <li class="clearable">
+        <span class="poster">Antonio</span> <span class="body">How are you?</span>
+      </li>
+      <li class="clearable">
+        <span class="poster">Antonio</span> <span class="body">Good, you?</span>
+      </li>
+...
+```
+
+Initially, we'll generate a username for the current user. We can store it in a
+`SessionVar`. `SessionVar`s in Lift are used to store things that should exist
+for the duration of a user's session. A user's session exists as long as Lift
+is aware of the user viewing a page related to that session. If Lift sees no
+activity from a given session after 20 minutes foonote:[This is configurable,
+of course. See `LiftRules.sessionInactivityTimeout`.], the session will be
+thrown away, as will the associated `SessionVar` values and related data.
+
+For now, let's look at adding the `SessionVar` to the `Chat` snippet:
+
+```
+...
+object username extends SessionVar[String]("username")
+
+object Chat {
+...
+```
+
+Here, we create a new `SessionVar`, whose default value will be “username” if it
+is accessed without having been set. We can change that to be random:
+
+```
+object username extends SessionVar[String]("User " + randomString(5))
+```
+
+We're using a Lift helper called `randomString`. We just pass it a length and
+it gives us back a random string of that length. This'll make sure that each
+user session has a (reasonably) unique username.
+
+Now, we need to store usernames alongside messages. Let's do that by making the
+messageEntries list contain a case class instance instead of a simple `String`:
+
+```
+...
+case class ChatMessage(poster: String, body: String)
+class Chat {
+  var messageEntries = List[ChatMessage]()
+
+  def messages = {
+    ClearClearable &
+    "li" #> messageEntries.map { entry =>
+      ".poster *" #> entry.poster &
+      ".body *" #> entry.body
+    }
+  }
+
+  def sendMessage = {
+    var message = ChatMessage("", "")
+
+    "#new-message" #> SHtml.text(message, { body: String => message = ChatMessage(username.is, body) }) &
```

Comment:
Some newbie questions that this line gave for me:

- What is this curly bracket thing? - is it an anonymous function definition in Scala? `text` expects a function as it second parameter, so maybe `String => message` is the function type. But `message` isn't a type, hmm. No, wait, aha - `body` is a self type in an anonymous class and the mutable variable `message` is set to a new ChatMessage instance - or? Having it on one line is really confusing. Do we need an anonymous class with a String self type?

- What is this method `is` on `username`? I pasted the code into a Lift project and realised it was simply a getter method on the username session variable to retrieve the value, duh. There's also a shadow method `get` that points to the `is` method. For a newbie tutorial, I would suggest to simply use the `get` method instead. For seasoned Lift users the `is` method is probably obvious but `get` is more generally understood when coming from the outside (as the audience of a tutorial would likely be).

Since `message` is now a ChatMessage, we need to say `message.body` for the first argument to `text`, no?

Could we instead say something like this?

```scala
"#new-message" #> SHtml.text(message.body, newMessage => { message = ChatMessage(username.get, newMessage) }) &
```

It would be great to describe the line in a similar manner as you so well described the previous lines in part 5:

```scala
    "#new-message" #> SHtml.text(message, message = _) &
    "type=submit" #> SHtml.submitButton(() => {
      messageEntries ::= message
    })
```
Once understanding passing NodeSes => NodeSeq functions around, css-selectors become more natural. So I think it is worth explaining examples of those thoroughly.